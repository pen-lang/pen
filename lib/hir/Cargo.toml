[package]
name = "hir"
version = "0.1.0"
authors = ["Yota Toyama <raviqqe@gmail.com>"]
edition = "2021"

[dependencies]
fnv = "1"
position = { path = "../position" }
serde = { version = "1", features = ["derive", "rc"] }

[dev-dependencies]
once_cell = "1"
pretty_assertions = "1"
