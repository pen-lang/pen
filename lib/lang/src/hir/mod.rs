pub mod analysis;
mod argument;
mod arithmetic_operation;
mod boolean;
mod boolean_operation;
mod call;
mod declaration;
mod definition;
mod else_branch;
mod equality_operation;
mod expression;
mod if_;
mod if_list;
mod if_type;
mod if_type_branch;
mod lambda;
mod let_;
mod list;
mod list_element;
mod module;
mod none;
mod not_operation;
mod number;
mod operation;
mod order_operation;
mod record_construction;
mod record_deconstruction;
mod record_element;
mod record_update;
mod string;
mod try_operation;
mod type_alias;
mod type_coercion;
mod type_definition;
mod variable;

pub use argument::*;
pub use arithmetic_operation::*;
pub use boolean::*;
pub use boolean_operation::*;
pub use call::*;
pub use declaration::*;
pub use definition::*;
pub use else_branch::*;
pub use equality_operation::*;
pub use expression::*;
pub use if_::*;
pub use if_list::*;
pub use if_type::*;
pub use if_type_branch::*;
pub use lambda::*;
pub use let_::*;
pub use list::*;
pub use list_element::*;
pub use module::*;
pub use none::*;
pub use not_operation::*;
pub use number::*;
pub use operation::*;
pub use order_operation::*;
pub use record_construction::*;
pub use record_deconstruction::*;
pub use record_element::*;
pub use record_update::*;
pub use string::*;
pub use try_operation::*;
pub use type_alias::*;
pub use type_coercion::*;
pub use type_definition::*;
pub use variable::*;
