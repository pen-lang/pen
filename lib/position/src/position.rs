use crate::format;
use serde::{Deserialize, Serialize};
use std::{
    cmp::Ordering,
    fmt::{self, Display, Formatter},
    hash::{Hash, Hasher},
    rc::Rc,
};

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Position(Rc<PositionInner>);

#[derive(Clone, Debug, Deserialize, Serialize)]
struct PositionInner {
    path: String,
    line_number: usize,
    column_number: usize,
    line: String,
}

impl Position {
    pub fn new(
        path: impl Into<String>,
        line_number: usize,
        column_number: usize,
        line: impl Into<String>,
    ) -> Self {
        Self(
            PositionInner {
                path: path.into(),
                line_number,
                column_number,
                line: line.into(),
            }
            .into(),
        )
    }

    pub fn path(&self) -> &str {
        &self.0.path
    }

    pub fn line_number(&self) -> usize {
        self.0.line_number
    }

    pub fn column_number(&self) -> usize {
        self.0.column_number
    }

    pub fn line(&self) -> &str {
        &self.0.line
    }
}

impl Display for Position {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        write!(formatter, "{}", format(self))
    }
}

impl Eq for Position {}

impl PartialEq for Position {
    fn eq(&self, _: &Self) -> bool {
        true
    }
}

impl Hash for Position {
    fn hash<H: Hasher>(&self, _: &mut H) {}
}

impl Ord for Position {
    fn cmp(&self, _: &Self) -> Ordering {
        Ordering::Equal
    }
}

impl PartialOrd for Position {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

#[cfg(test)]
mod tests {
    use super::Position;

    #[test]
    fn display() {
        assert_eq!(
            format!("{}", Position::new("file", 1, 1, "x")),
            "file\n1:1:\tx\n    \t^"
        );

        assert_eq!(
            format!("{}", Position::new("file", 1, 2, " x")),
            "file\n1:2:\t x\n    \t ^"
        );
    }
}
