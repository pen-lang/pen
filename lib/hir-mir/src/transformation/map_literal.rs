use super::{collection_type, map_context};
use crate::{context::CompileContext, CompileError};
use hir::{
    ir::*,
    types::{self, Type},
};
use position::Position;

pub fn transform(context: &CompileContext, map: &Map) -> Result<Expression, CompileError> {
    transform_map(
        context,
        map.key_type(),
        map.value_type(),
        map.elements(),
        map.position(),
    )
}

fn transform_map(
    context: &CompileContext,
    key_type: &Type,
    value_type: &Type,
    elements: &[MapElement],
    position: &Position,
) -> Result<Expression, CompileError> {
    let configuration = &context.configuration()?.map_type;
    let map_context_type = collection_type::transform_map_context(context, position)?;
    let any_map_type = collection_type::transform_map(context, position)?;
    let map_context = map_context::transform(context, key_type, value_type, position)?;

    Ok(match elements {
        [] => Call::new(
            Some(types::Function::new(vec![], any_map_type, position.clone()).into()),
            Variable::new(&configuration.empty_function_name, position.clone()),
            vec![],
            position.clone(),
        )
        .into(),
        // TODO Optimize cases where only a single element of MapElement::Map
        // exists when we pass context functions dynamically.
        [.., element] => {
            let rest_expression = transform_map(
                context,
                key_type,
                value_type,
                &elements[..elements.len() - 1],
                position,
            )?;

            match element {
                MapElement::Insertion(entry) => Call::new(
                    Some(
                        types::Function::new(
                            vec![
                                map_context_type,
                                any_map_type.clone(),
                                types::Any::new(position.clone()).into(),
                                types::Any::new(position.clone()).into(),
                            ],
                            any_map_type,
                            position.clone(),
                        )
                        .into(),
                    ),
                    Variable::new(&configuration.set_function_name, position.clone()),
                    vec![
                        map_context,
                        rest_expression,
                        TypeCoercion::new(
                            key_type.clone(),
                            types::Any::new(position.clone()),
                            entry.key().clone(),
                            position.clone(),
                        )
                        .into(),
                        TypeCoercion::new(
                            value_type.clone(),
                            types::Any::new(position.clone()),
                            entry.value().clone(),
                            position.clone(),
                        )
                        .into(),
                    ],
                    position.clone(),
                )
                .into(),
                MapElement::Map(expression) => Call::new(
                    Some(
                        types::Function::new(
                            vec![map_context_type, any_map_type.clone(), any_map_type.clone()],
                            any_map_type,
                            position.clone(),
                        )
                        .into(),
                    ),
                    Variable::new(&configuration.merge_function_name, position.clone()),
                    vec![map_context, expression.clone(), rest_expression],
                    position.clone(),
                )
                .into(),
                MapElement::Removal(key) => Call::new(
                    Some(
                        types::Function::new(
                            vec![
                                map_context_type,
                                any_map_type.clone(),
                                types::Any::new(position.clone()).into(),
                            ],
                            any_map_type,
                            position.clone(),
                        )
                        .into(),
                    ),
                    Variable::new(&configuration.delete_function_name, position.clone()),
                    vec![
                        map_context,
                        rest_expression,
                        TypeCoercion::new(
                            key_type.clone(),
                            types::Any::new(position.clone()),
                            key.clone(),
                            position.clone(),
                        )
                        .into(),
                    ],
                    position.clone(),
                )
                .into(),
            }
        }
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use position::test::PositionFake;

    #[test]
    fn transform_empty_map() {
        insta::assert_debug_snapshot!(transform(
            &CompileContext::dummy(Default::default(), Default::default()),
            &Map::new(
                types::None::new(Position::fake()),
                types::None::new(Position::fake()),
                vec![],
                Position::fake()
            ),
        ));
    }

    #[test]
    fn transform_empty_map_with_function_value() {
        insta::assert_debug_snapshot!(transform(
            &CompileContext::dummy(Default::default(), Default::default()),
            &Map::new(
                types::None::new(Position::fake()),
                types::Function::new(vec![], types::None::new(Position::fake()), Position::fake()),
                vec![],
                Position::fake()
            ),
        ));
    }

    #[test]
    fn transform_map_with_entry() {
        insta::assert_debug_snapshot!(transform(
            &CompileContext::dummy(Default::default(), Default::default()),
            &Map::new(
                types::None::new(Position::fake()),
                types::None::new(Position::fake()),
                vec![MapEntry::new(
                    None::new(Position::fake()),
                    None::new(Position::fake()),
                    Position::fake()
                )
                .into()],
                Position::fake()
            ),
        ));
    }
}
