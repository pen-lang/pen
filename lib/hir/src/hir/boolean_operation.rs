use super::expression::Expression;
use position::Position;
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum BooleanOperator {
    And,
    Or,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BooleanOperation {
    operator: BooleanOperator,
    lhs: Arc<Expression>,
    rhs: Arc<Expression>,
    position: Position,
}

impl BooleanOperation {
    pub fn new(
        operator: BooleanOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            operator,
            lhs: lhs.into().into(),
            rhs: rhs.into().into(),
            position,
        }
    }

    pub fn operator(&self) -> BooleanOperator {
        self.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.rhs
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
