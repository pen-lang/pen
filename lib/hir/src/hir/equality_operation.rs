use super::expression::Expression;
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum EqualityOperator {
    Equal,
    NotEqual,
}

#[derive(Clone, Debug, PartialEq)]
pub struct EqualityOperation {
    type_: Option<Type>,
    operator: EqualityOperator,
    lhs: Arc<Expression>,
    rhs: Arc<Expression>,
    position: Position,
}

impl EqualityOperation {
    pub fn new(
        type_: Option<Type>,
        operator: EqualityOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            type_,
            operator,
            lhs: Arc::new(lhs.into()),
            rhs: Arc::new(rhs.into()),
            position,
        }
    }

    pub fn type_(&self) -> Option<&Type> {
        self.type_.as_ref()
    }

    pub fn operator(&self) -> EqualityOperator {
        self.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.rhs
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
