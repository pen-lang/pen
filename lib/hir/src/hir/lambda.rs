use super::{Argument, Expression};
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct Lambda {
    arguments: Vec<Argument>,
    result_type: Type,
    body: Arc<Expression>,
    position: Position,
}

impl Lambda {
    pub fn new(
        arguments: Vec<Argument>,
        result_type: impl Into<Type>,
        body: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            arguments,
            result_type: result_type.into(),
            body: body.into().into(),
            position,
        }
    }

    pub fn arguments(&self) -> &[Argument] {
        &self.arguments
    }

    pub fn result_type(&self) -> &Type {
        &self.result_type
    }

    pub fn body(&self) -> &Expression {
        &self.body
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
