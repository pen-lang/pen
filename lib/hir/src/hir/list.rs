use super::list_element::ListElement;
use crate::types::Type;
use position::Position;

#[derive(Clone, Debug, PartialEq)]
pub struct List {
    type_: Type,
    elements: Vec<ListElement>,
    position: Position,
}

impl List {
    pub fn new(type_: impl Into<Type>, elements: Vec<ListElement>, position: Position) -> Self {
        Self {
            type_: type_.into(),
            elements,
            position,
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn elements(&self) -> &[ListElement] {
        &self.elements
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
