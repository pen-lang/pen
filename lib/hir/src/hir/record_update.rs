use super::{expression::Expression, RecordElement};
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct RecordUpdate {
    type_: Type,
    record: Arc<Expression>,
    elements: Vec<RecordElement>,
    position: Position,
}

impl RecordUpdate {
    pub fn new(
        type_: impl Into<Type>,
        record: impl Into<Expression>,
        elements: Vec<RecordElement>,
        position: Position,
    ) -> Self {
        Self {
            type_: type_.into(),
            record: Arc::new(record.into()),
            elements,
            position,
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn record(&self) -> &Expression {
        &self.record
    }

    pub fn elements(&self) -> &[RecordElement] {
        &self.elements
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
