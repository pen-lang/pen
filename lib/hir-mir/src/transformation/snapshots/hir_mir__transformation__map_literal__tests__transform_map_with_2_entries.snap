---
source: lib/hir-mir/src/transformation/map_literal.rs
expression: "transform(&Context::dummy(Default::default(), Default::default()),\n    &Map::new(types::None::new(Position::fake()),\n            types::None::new(Position::fake()),\n            vec![MapEntry ::\n                new(Number :: new(1.0, Position :: fake()), None ::\n                new(Position :: fake()), Position :: fake()).into(), MapEntry\n                ::\n                new(Number :: new(2.0, Position :: fake()), None ::\n                new(Position :: fake()), Position :: fake()).into()],\n            Position::fake()))"
---
Ok(
    Let(
        Let {
            name: Some(
                "$ctx",
            ),
            type_: Some(
                Reference(
                    Reference(
                        ReferenceInner {
                            name: "mapContext",
                            position: Position(
                                PositionInner {
                                    path: "",
                                    line_number: 1,
                                    column_number: 1,
                                    line: "",
                                },
                            ),
                        },
                    ),
                ),
            ),
            bound_expression: Call(
                Call {
                    function_type: Some(
                        Function(
                            Function(
                                FunctionInner {
                                    arguments: [],
                                    result: Reference(
                                        Reference(
                                            ReferenceInner {
                                                name: "mapContext",
                                                position: Position(
                                                    PositionInner {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                ),
                                            },
                                        ),
                                    ),
                                    position: Position(
                                        PositionInner {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    ),
                                },
                            ),
                        ),
                    ),
                    function: Variable(
                        Variable {
                            name: "hir:map:context:829f738c8f9ee2f4",
                            position: Position(
                                PositionInner {
                                    path: "",
                                    line_number: 1,
                                    column_number: 1,
                                    line: "",
                                },
                            ),
                        },
                    ),
                    arguments: [],
                    position: Position(
                        PositionInner {
                            path: "",
                            line_number: 1,
                            column_number: 1,
                            line: "",
                        },
                    ),
                },
            ),
            expression: Call(
                Call {
                    function_type: Some(
                        Function(
                            Function(
                                FunctionInner {
                                    arguments: [
                                        Reference(
                                            Reference(
                                                ReferenceInner {
                                                    name: "mapContext",
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                        ),
                                        Reference(
                                            Reference(
                                                ReferenceInner {
                                                    name: "genericMap",
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                        ),
                                        Any(
                                            Any {
                                                position: Position(
                                                    PositionInner {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                ),
                                            },
                                        ),
                                        Any(
                                            Any {
                                                position: Position(
                                                    PositionInner {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                ),
                                            },
                                        ),
                                    ],
                                    result: Reference(
                                        Reference(
                                            ReferenceInner {
                                                name: "genericMap",
                                                position: Position(
                                                    PositionInner {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                ),
                                            },
                                        ),
                                    ),
                                    position: Position(
                                        PositionInner {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    ),
                                },
                            ),
                        ),
                    ),
                    function: Variable(
                        Variable {
                            name: "setMap",
                            position: Position(
                                PositionInner {
                                    path: "",
                                    line_number: 1,
                                    column_number: 1,
                                    line: "",
                                },
                            ),
                        },
                    ),
                    arguments: [
                        Variable(
                            Variable {
                                name: "$ctx",
                                position: Position(
                                    PositionInner {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                ),
                            },
                        ),
                        Call(
                            Call {
                                function_type: Some(
                                    Function(
                                        Function(
                                            FunctionInner {
                                                arguments: [
                                                    Reference(
                                                        Reference(
                                                            ReferenceInner {
                                                                name: "mapContext",
                                                                position: Position(
                                                                    PositionInner {
                                                                        path: "",
                                                                        line_number: 1,
                                                                        column_number: 1,
                                                                        line: "",
                                                                    },
                                                                ),
                                                            },
                                                        ),
                                                    ),
                                                    Reference(
                                                        Reference(
                                                            ReferenceInner {
                                                                name: "genericMap",
                                                                position: Position(
                                                                    PositionInner {
                                                                        path: "",
                                                                        line_number: 1,
                                                                        column_number: 1,
                                                                        line: "",
                                                                    },
                                                                ),
                                                            },
                                                        ),
                                                    ),
                                                    Any(
                                                        Any {
                                                            position: Position(
                                                                PositionInner {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            ),
                                                        },
                                                    ),
                                                    Any(
                                                        Any {
                                                            position: Position(
                                                                PositionInner {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            ),
                                                        },
                                                    ),
                                                ],
                                                result: Reference(
                                                    Reference(
                                                        ReferenceInner {
                                                            name: "genericMap",
                                                            position: Position(
                                                                PositionInner {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            ),
                                                        },
                                                    ),
                                                ),
                                                position: Position(
                                                    PositionInner {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                ),
                                            },
                                        ),
                                    ),
                                ),
                                function: Variable(
                                    Variable {
                                        name: "setMap",
                                        position: Position(
                                            PositionInner {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        ),
                                    },
                                ),
                                arguments: [
                                    Variable(
                                        Variable {
                                            name: "$ctx",
                                            position: Position(
                                                PositionInner {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            ),
                                        },
                                    ),
                                    Call(
                                        Call {
                                            function_type: Some(
                                                Function(
                                                    Function(
                                                        FunctionInner {
                                                            arguments: [],
                                                            result: Reference(
                                                                Reference(
                                                                    ReferenceInner {
                                                                        name: "genericMap",
                                                                        position: Position(
                                                                            PositionInner {
                                                                                path: "",
                                                                                line_number: 1,
                                                                                column_number: 1,
                                                                                line: "",
                                                                            },
                                                                        ),
                                                                    },
                                                                ),
                                                            ),
                                                            position: Position(
                                                                PositionInner {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            ),
                                                        },
                                                    ),
                                                ),
                                            ),
                                            function: Variable(
                                                Variable {
                                                    name: "emptyMap",
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                            arguments: [],
                                            position: Position(
                                                PositionInner {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            ),
                                        },
                                    ),
                                    TypeCoercion(
                                        TypeCoercion {
                                            from: None(
                                                None {
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                            to: Any(
                                                Any {
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                            argument: Number(
                                                Number {
                                                    value: 1.0,
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                            position: Position(
                                                PositionInner {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            ),
                                        },
                                    ),
                                    TypeCoercion(
                                        TypeCoercion {
                                            from: None(
                                                None {
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                            to: Any(
                                                Any {
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                            argument: None(
                                                None {
                                                    position: Position(
                                                        PositionInner {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    ),
                                                },
                                            ),
                                            position: Position(
                                                PositionInner {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            ),
                                        },
                                    ),
                                ],
                                position: Position(
                                    PositionInner {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                ),
                            },
                        ),
                        TypeCoercion(
                            TypeCoercion {
                                from: None(
                                    None {
                                        position: Position(
                                            PositionInner {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        ),
                                    },
                                ),
                                to: Any(
                                    Any {
                                        position: Position(
                                            PositionInner {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        ),
                                    },
                                ),
                                argument: Number(
                                    Number {
                                        value: 2.0,
                                        position: Position(
                                            PositionInner {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        ),
                                    },
                                ),
                                position: Position(
                                    PositionInner {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                ),
                            },
                        ),
                        TypeCoercion(
                            TypeCoercion {
                                from: None(
                                    None {
                                        position: Position(
                                            PositionInner {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        ),
                                    },
                                ),
                                to: Any(
                                    Any {
                                        position: Position(
                                            PositionInner {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        ),
                                    },
                                ),
                                argument: None(
                                    None {
                                        position: Position(
                                            PositionInner {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        ),
                                    },
                                ),
                                position: Position(
                                    PositionInner {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                ),
                            },
                        ),
                    ],
                    position: Position(
                        PositionInner {
                            path: "",
                            line_number: 1,
                            column_number: 1,
                            line: "",
                        },
                    ),
                },
            ),
            position: Position(
                PositionInner {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            ),
        },
    ),
)
